besu/src/main/java/org/hyperledger/besu/cli/subcommands/operator/RestoreState.java

```diff
@@ -0,0 +1,348 @@
+        final BlockHeaderFunctions functions = new MainnetBlockHeaderFunctions();
+
+        final BlockHeader header =
+            BlockHeader.readFrom(
+                new BytesValueRLPInput(Bytes.wrap(headerEntry), false, true), functions);
+        final BlockBody body =
+            BlockBody.readFrom(
+                new BytesValueRLPInput(Bytes.wrap(bodyEntry), false, true), functions);
+        final RLPInput receiptsRlp = new BytesValueRLPInput(Bytes.wrap(receiptEntry), false, true);
+        final int receiptsCount = receiptsRlp.enterList();
+        final List<TransactionReceipt> receipts = new ArrayList<>(receiptsCount);
+        for (int j = 0; j < receiptsCount; j++) {
+          receipts.add(TransactionReceipt.readFrom(receiptsRlp, true));
+        }
+        receiptsRlp.leaveList();
+
+        blockchain.appendBlock(new Block(header, body), receipts);
+      }
+    }
+    LOG.info("Chain data loaded");
+  }
+
+  @SuppressWarnings("UnusedVariable")
+  private void restoreAccounts() throws IOException {
+    newWorldStateUpdater();
+    int storageBranchCount = 0;
+    int storageExtensionCount = 0;
+    int storageLeafCount = 0;
+
+    final PersistVisitor<Bytes> accountPersistVisitor =
+        new PersistVisitor<>(this::updateAccountState);
+    Node<Bytes> root = accountPersistVisitor.initialRoot();
+
+    try (final RollingFileReader reader =
+        new RollingFileReader(this::accountFileName, compressed)) {
+      for (int i = 0; i < accountCount; i++) {
+        if (i % 100000 == 0) {
+          LOG.info("Loading account data {} / {}", i, accountCount);
+        }
+        final byte[] accountEntry = reader.readBytes();
+        final BytesValueRLPInput accountInput =
+            new BytesValueRLPInput(Bytes.of(accountEntry), false, true);
+        final int length = accountInput.enterList();
+        if (length != 3) {
+          throw new RuntimeException("Unexpected account length " + length);
+        }
+        final Bytes32 trieKey = accountInput.readBytes32();
+        final Bytes accountRlp = accountInput.readBytes();
+        final Bytes code = accountInput.readBytes();
+
+        final StateTrieAccountValue trieAccount =
+            StateTrieAccountValue.readFrom(new BytesValueRLPInput(accountRlp, false, true));
+        if (!trieAccount.getCodeHash().equals(Hash.hash(code))) {
+          throw new RuntimeException("Code hash doesn't match");
+        }
+        if (code.size() > 0) {
+          updateCode(code);
+        }
+
+        final RestoreVisitor<Bytes> accountTrieWriteVisitor =
+            new RestoreVisitor<>(t -> t, accountRlp, accountPersistVisitor);
+
+        root = root.accept(accountTrieWriteVisitor, bytesToPath(trieKey));
+
+        final PersistVisitor<Bytes> storagePersistVisitor =
+            new PersistVisitor<>(this::updateAccountStorage);
+        Node<Bytes> storageRoot = storagePersistVisitor.initialRoot();
+
+        while (true) {
```

Comment:
Is it not possible to have something like while(len!=0) instead of while (true) ?

Reply:
Because we don't know a-priori what the length of the list of leaf nodes will be.  It's too large for memory (I tried) and reading it twice imposes a huge DB hit.  So I null terminate the list.

Making it a len check would involve duplicating the code to read the trie value.